components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt

```diff
@@ -236,28 +239,52 @@ class FennecMigrator private constructor(
      * @return A deferred [MigrationResults], describing which migrations were performed and if they succeeded.
      */
     fun migrateAsync(): Deferred<MigrationResults> = synchronized(migrationLock) {
+        val migrationsToRun = getMigrationsToRun()
+
+        // Short-circuit if there's nothing to do.
+        if (migrationsToRun.isEmpty()) {
+            logger.debug("No migrationz to run.")
+            val result = CompletableDeferred<MigrationResults>()
+            result.complete(emptyMap())
+            return result
+        }
+
+        return runMigrationsAsync(migrationsToRun)
+    }
+
+    /**
+     * Returns true if there are migrations to run for this installation.
+     */
+    private fun hasMigrationsToRun(): Boolean {
+        return getMigrationsToRun().isNotEmpty()
+    }
+
+    /**
+     * Starts the provided [AbstractMigrationService] implementation if there are migrations to be
+     * run for this installation.
+     */
+    fun <T> startMigrationServiceIfNeeded(service: Class<T>) where T : AbstractMigrationService {
+        if (hasMigrationsToRun()) {
+            logger.debug("Has migrations to run. Starting service.")
+            context.startService(Intent(context, service))
+        } else {
+            logger.debug("No migrations to run. Not starting service.")
+        }
+    }
+
+    private fun getMigrationsToRun(): List<VersionedMigration> {
```

Comment:
I'm a little worried that this is now not protected by the `migrationLock`; we _shouldn't_ have overlapping migrations running, of course...

Reply:
Yeah, overlapping migrations could cause a bunch of other issues.

It's private and still protected by the lock when called from `migrateAsync()`. I think I'll add the same protection to `startMigrationServiceIfNeeded()`.

Reply:
Or maybe not! If a migration is currently running then we do not want to block here - because that can take a while. 🤔

Reply:
I think I want to land as-is. Once we have #4872 it will be easier to communicate the current state.